--- httpd-2.4.54.orig/support/suexec.c	2020-07-03 22:53:37.000000000 +0300
+++ httpd-2.4.54/support/suexec.c	2022-12-23 11:57:42.799970269 +0200
@@ -339,6 +339,7 @@ int main(int argc, char *argv[])
 #ifdef AP_USERDIR_SUFFIX
         fprintf(stderr, " -D AP_USERDIR_SUFFIX=\"%s\"\n", AP_USERDIR_SUFFIX);
 #endif
+        fprintf(stderr, " -D AP_SAFE_DIRECTORY=\"%s\"\n", AP_SAFE_DIRECTORY);
         exit(0);
     }
     /*
@@ -529,6 +530,23 @@ int main(int argc, char *argv[])
         exit(111);
     }
 
+    /* Check for safe directory existence */
+    char safe_dr[AP_MAXPATH];
+    int is_safe_dir_present = 0;
+    struct stat safe_dir_info;
+    if (((lstat(AP_SAFE_DIRECTORY, &safe_dir_info)) == 0) && (S_ISDIR(safe_dir_info.st_mode))) {
+    	is_safe_dir_present = 1;
+    }
+
+    if(is_safe_dir_present){
+    	if (((chdir(AP_SAFE_DIRECTORY)) != 0) ||
+    		((getcwd(safe_dr, AP_MAXPATH)) == NULL) ||
+    		((chdir(cwd)) != 0)) {
+    		log_err("cannot get safe directory information (%s)\n", AP_SAFE_DIRECTORY);
+    		exit(200);
+    	}
+    }
+
     if (userdir) {
         if (((chdir(target_homedir)) != 0) ||
             ((chdir(AP_USERDIR_SUFFIX)) != 0) ||
@@ -547,9 +565,21 @@ int main(int argc, char *argv[])
         }
     }
 
-    if ((strncmp(cwd, dwd, strlen(dwd))) != 0) {
-        log_err("command not in docroot (%s/%s)\n", cwd, cmd);
-        exit(114);
+    int safe_work = 0;
+    if(is_safe_dir_present){
+    	if ((strncmp(cwd, safe_dr, strlen(safe_dr))) != 0){
+    		if ((strncmp(cwd, dwd, strlen(dwd))) != 0) {
+    		  log_err("command not in docroot (%s/%s)\n", cwd, cmd);
+    		  exit(114);
+    		}
+    	} else {
+    		safe_work = 1;
+    	}
+    } else {
+    	if ((strncmp(cwd, dwd, strlen(dwd))) != 0) {
+    	  log_err("command not in docroot (%s/%s)\n", cwd, cmd);
+    	  exit(114);
+    	}
     }
 
     /*
@@ -596,6 +626,7 @@ int main(int argc, char *argv[])
      * Error out if the target name/group is different from
      * the name/group of the cwd or the program.
      */
+    if (!safe_work){
     if ((uid != dir_info.st_uid) ||
         (gid != dir_info.st_gid) ||
         (uid != prg_info.st_uid) ||
@@ -607,6 +638,7 @@ int main(int argc, char *argv[])
                 (unsigned long)prg_info.st_uid, (unsigned long)prg_info.st_gid);
         exit(120);
     }
+    }
     /*
      * Error out if the program is not executable for the user.
      * Otherwise, she won't find any error in the logs except for
--- httpd-2.4.54.orig/support/suexec.h	2011-09-23 16:38:09.000000000 +0300
+++ httpd-2.4.54/support/suexec.h	2022-12-23 12:05:09.891977168 +0200
@@ -105,5 +105,12 @@
 #ifndef AP_SAFE_PATH
 #define AP_SAFE_PATH "/usr/local/bin:/usr/bin:/bin"
 #endif
+/*
+ * SAFE_DIRECTORY -- Define a safe directory to execute files from.
+ *
+ */
+#ifndef AP_SAFE_DIRECTORY
+#define AP_SAFE_DIRECTORY "/usr/local/safe-bin"
+#endif
 
 #endif /* _SUEXEC_H */
